statedb/statedb.go

```diff
@@ -1,35 +1,720 @@
+	tree, err := tree.New(&readOnlyWriteTx{txTree},
+		tree.Options{DB: subDB(db, subKeyTree), MaxLevels: cfg.maxLevels, HashFunc: cfg.hashFunc})
+	if err == ErrReadOnly {
+		return nil, ErrEmptyTree
+	} else if err != nil {
+		return nil, err
+	}
+	tree, err = tree.FromRoot(root)
+	if err != nil {
+		return nil, err
+	}
+	return &TreeView{
+		db:   db,
+		tree: tree,
+		cfg:  mainTreeCfg,
+	}, nil
+}
+
+// SubTreeSingle returns a TreeView of a singleton SubTree whose root is stored
+// in the leaf with `cfg.Key()`, and is parametrized by `cfg`.
+func (v *TreeView) SubTreeSingle(c *SubTreeSingleConfig) (*TreeView, error) {
+	return v.subTree(c.treeConfig())
+}
+
+// SubTree returns a TreeView of a non-singleton SubTree whose root is stored
+// in the leaf with `key`, and is parametrized by `cfg`.
+func (v *TreeView) SubTree(key []byte, c *SubTreeConfig) (*TreeView, error) {
+	return v.subTree(c.treeConfig(key))
+}
+
+// TreeUpdate is an opened tree that can be updated.  All updates are stored in
+// an internal transaction (shared with all the opened subTrees) and will only
+// be commited with the TreeTx.Commit function is called.
+type TreeUpdate struct {
+	tx        db.WriteTx
+	txTree    db.WriteTx
+	dirtyTree bool
+	tree      *tree.Tree
+	openSubs  map[string]*TreeUpdate
+	cfg       *treeConfig
+}
+
+// Get returns the value at key in this tree.  `key` is the path of the leaf,
+// and the returned value is the leaf's value.
+func (u *TreeUpdate) Get(key []byte) ([]byte, error) {
+	return u.tree.Get(u.txTree, key)
+}
+
+// Iterate iterates over all nodes of this tree.
+func (u *TreeUpdate) Iterate(callback func(key, value []byte) bool) error {
+	return u.tree.Iterate(u.txTree, callback)
+}
+
+// Root returns the root of the tree, which cryptographically summarises the
+// state of the tree.
+func (u *TreeUpdate) Root() ([]byte, error) {
+	return u.tree.Root(u.txTree)
+}
+
+// Size returns the number of leafs (key-values) that this tree contains.
+func (u *TreeUpdate) Size() (uint64, error) {
+	// NOTE: Tree.Size is currently unimplemented
+	return u.tree.Size(u.txTree), nil
+}
+
+// GenProof generates a proof of existence of the given key for this tree.  The
+// returned values are the leaf value and the proof itself.
+func (u *TreeUpdate) GenProof(key []byte) ([]byte, []byte, error) {
+	return u.tree.GenProof(u.txTree, key)
+}
+
+// Unimplemented because arbo.Tree.Dump doesn't take db.ReadTx as input.
+// func (u *TreeUpdate) Dump() ([]byte, error) {
+// 	panic("TODO")
+// }
+
+// NoState returns a key-value database associated with this tree that doesn't
+// affect the cryptographic integrity of the StateDB.  Writing to this database
+// won't change the StateDB.Root.
+func (u *TreeUpdate) NoState() Updater {
+	return subWriteTx(u.tx, subKeyNoState)
+}
+
+// Add a new key-value to this tree.  `key` is the path of the leaf, and
+// `value` is the content of the leaf.
+func (u *TreeUpdate) Add(key, value []byte) error {
+	u.dirtyTree = true
```

Comment:
if these methods are meant to be safe for concurrent use (e.g. calling Add/Set from many goroutines), you could use sync/atomic for this field

Reply:
The TreeUpdate and TreeTx are not safe for concurrent use (it's not in our use case); so I've added that to the documentation of the types.